### PR TITLE
[5.8] Add the ability to register custom DBAL types in the schema builder.

### PR DESCRIPTION
**Pull request by jensdenies**
### Overview
This pull request aims to resolve issue #8840. In addition to that, it allows you to easily add custom, platform specific DBAL types to the framework in the future.

### Quick Summary
When changing an existing column to tinyint, it will produce the following error:

> DBALException in DBALException.php line 228:
Unknown column type "tinyinteger" requested. Any Doctrine type that you use has to be registered with \Doctrine\DBAL\Types\Type::addType()

### Reproduction
In order to reproduce this error, you'll need to run two migrations. One should create a column and the other should should change the column:

```php
// The first migration.
Schema::create('test_migration', function($table) {
    $table->char('test_column');
});
```
```php
// The second migration.
Schema::table('test_migration', function($table) {
    $table->tinyInteger('test_column')->change();
});
```

### Proposed Solution
In order to solve this problem, i created a "registerCustomDBALType" method in the "Builder" class. This way the platform specific builder can call this method to register their own custom type. In addition to that, the user can register their own type in their project by calling the "registerCustomDBALType" on the "Schema" facade.

### CODE REVIEW

**Comment by staudenmeir** [comment]
We should also register the type on other databases, at least on SQL Server (which has a native `tinyint` type). Maybe even on PostgreSQL and SQLite.

Why are you overriding `convertToPHPValue()` and `convertToDatabaseValue()` without changing the implementation?

Please add a test that checks the generated SQL.

**Comment by JacksonIV** [comment]
@staudenmeir I removed the unnecessary overrides, still working on the rest though.

File: src/Illuminate/Database/Schema/Builder.php [hunk 1]
```diff
@@ -317,4 +318,28 @@ public function blueprintResolver(Closure $resolver)
     {
         $this->resolver = $resolver;
     }
+
+    /**
+     * Register your own Doctrine mapping type.
+     *
+     * @param  string  $class
+     * @param  string  $name
+     * @param  string $type
```
Comment (driesvints) [inline src/Illuminate/Database/Schema/Builder.php]: Add a space here
```suggestion
     * @param  string  $type
```

File: src/Illuminate/Database/Schema/MySqlBuilder.php [hunk 2]
```diff
@@ -2,8 +2,25 @@
...
+use Illuminate\Database\Connection;
+use Illuminate\Database\Schema\Types\TinyInteger;
+
 class MySqlBuilder extends Builder
 {
+    /**
+     * MySqlBuilder constructor.
+     *
+     * @param  Connection  $connection
+     *
```
Comment (driesvints) [inline src/Illuminate/Database/Schema/MySqlBuilder.php]: Remove this line
```suggestion
```

File: src/Illuminate/Database/Schema/MySqlBuilder.php [hunk 3]
```diff
@@ -2,8 +2,25 @@
...
 
+use Illuminate\Database\Connection;
+use Illuminate\Database\Schema\Types\TinyInteger;
+
 class MySqlBuilder extends Builder
 {
+    /**
+     * MySqlBuilder constructor.
+     *
+     * @param  Connection  $connection
```
Comment (driesvints) [inline src/Illuminate/Database/Schema/MySqlBuilder.php]: Use FQN
```suggestion
     * @param  \Illuminate\Database\Connection  $connection
```

File: src/Illuminate/Database/Schema/Types/TinyInteger.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     *
+     * @var string
+     */
+    const NAME = 'tinyinteger';
+
+    /**
+     * Gets the SQL declaration snippet for a field of this type.
+     *
+     * @param  array  $fieldDeclaration
+     * @param  AbstractPlatform  $platform
```
Comment (driesvints) [inline src/Illuminate/Database/Schema/Types/TinyInteger.php]: Use FQN
```suggestion
     * @param  \Doctrine\DBAL\Platforms\AbstractPlatform  $platform
```

**Review by driesvints — CHANGES_REQUESTED** [review]

**Comment by JacksonIV** [comment]
@driesvints I made the requested changes.

**Comment by SturmB** [comment]
Hey, is there a chance someone could explain how to use this to get the dreaded "TINYINT" issue to disappear? I'm still learning and don't know how to use `registerCustomDBALType` properly (like where to call it, for example).

**Comment by driesvints** [comment]
@SturmB this was partially reverted again in https://github.com/laravel/framework/issues/28282 because of the following issue: https://github.com/laravel/framework/issues/28282

**Comment by JacksonIV** [comment]
@SturmB @driesvints I just want to let you know that i’m still working on a solution to fix the problems with TINYINT, as well as other column types.

**Pull request merged**